### PR TITLE
Add test cases for port auto negotiation feature

### PR DESCRIPTION
**Pull request by Junchao-Mellanox**
### Description of PR

Summary:
Fixes # (issue)

### Type of change

- [ ] Bug fix
- [ ] Testbed and Framework(new/improvement)
- [ ] Test case(new/improvement)

### CODE REVIEW

File: tests/platform_tests/mellanox/check_hw_mgmt_service.py [hunk 1]
```diff
@@ -13,10 +13,6 @@ def check_hw_management_service(dut):
     assert hw_mgmt_service_state["ActiveState"] == "active", "The hw-management service is not active"
     assert hw_mgmt_service_state["SubState"] == "exited", "The hw-management service is not exited"
 
-    logging.info("Check thermal control status")
```
Comment (keboliu) [inline tests/platform_tests/mellanox/check_hw_mgmt_service.py:16]: I remember we have another PR to handle this? suggest not to add this in this PR.

File: tests/platform_tests/test_auto_negotiation.py [hunk 2]
```diff
@@ -0,0 +1,356 @@
...
+    Args:
+        duthost: DUT object
+
+    Returns:
+        object: A helper class or instance
+    """
+    asic_type = duthost.facts["asic_type"]
+
+    if asic_type == "mellanox":
+        return MlnxCableSupportedSpeedsHelper
```
Comment (keboliu) [inline tests/platform_tests/test_auto_negotiation.py:307]: Is this also support 3rd party cable installed on Mellanox switch?

File: tests/platform_tests/test_auto_negotiation.py [hunk 3]
```diff
@@ -0,0 +1,356 @@
...
+            success = fanout.set_auto_negotiation_mode(fanout_port, False)
+            if not success:
+                # Fanout does not support set auto negotiation mode for this port
+                logger.info('Ignore port {} due to fanout port {} does not support setting auto-neg mode'.format(dut_port, fanout_port))
+                continue
+
+            logger.info('Trying to get a common supported speeds set among dut port, fanout port and cable')
+            supported_speeds = get_supported_speeds_for_port(duthost, dut_port, fanout, fanout_port)
+            if not supported_speeds:
+                continue
```
Comment (keboliu) [inline tests/platform_tests/test_auto_negotiation.py:273]: print out a log to indicate skip one port?

File: tests/platform_tests/test_auto_negotiation.py [hunk 4]
```diff
@@ -0,0 +1,356 @@
...
+    global cadidate_test_ports
+    fanout_original_port_states = {}
+    dutname, portname = decode_dut_port_name(enum_dut_portname_module_fixture)
+    logger.info('Collecting existing port configuration for DUT and fanout...')
+    for duthost in duthosts:
+        if dutname == 'unknown' or dutname == duthost.hostname:
+            all_ports = build_test_candidates(duthost, fanouthosts, portname)
+            # Test all ports takes too much time (sometimes more than an hour), 
+            # so we choose 3 ports randomly as the cadidates ports
+            cadidate_test_ports[duthost] = random.sample(all_ports, 3)
```
Comment (keboliu) [inline tests/platform_tests/test_auto_negotiation.py:50]: Will it happen in a high possibility that all the 3 ports are not testable? means fanout or cable to support AN?

File: tests/platform_tests/test_auto_negotiation.py [hunk 5]
```diff
@@ -0,0 +1,356 @@
...
+                continue
+
+            # Advertise all supported speeds in fanout port
+            success = fanout.set_speed(fanout_port, None)
+            if not success:
+                # Fanout does not support set advertise speeds for this port
+                logger.info('Ignore port {} due to fanout port {} does not support setting advertised speeds'.format(dut_port, fanout_port))
+                continue
+
+            logger.info('Trying to get a common supported speeds set among dut port, fanout port and cable')
```
Comment (keboliu) [inline tests/platform_tests/test_auto_negotiation.py:232]: speeds --> speed

File: tests/platform_tests/test_auto_negotiation.py (hunk 4 above)
Comment (Junchao-Mellanox) [reply tests/platform_tests/test_auto_negotiation.py:50]: I never met it during my test.

File: tests/platform_tests/test_auto_negotiation.py (hunk 2 above)
Comment (Junchao-Mellanox) [reply tests/platform_tests/test_auto_negotiation.py:307]: Yes, it is supported according to my test.

**Review by keboliu — APPROVED** [review]

**Pull request closed without merging**